File: src/tools/activate_skill.ts
import type { ToolDef, ToolResult, ToolExecutionContext } from "./registry.js";
import { register } from "./registry.js";
import { getSkillBody } from "#src/skills/index.js";

export const activateSkillTool: ToolDef = {
  name: "ActivateSkill",
  description:
    "Loads the full instructions of a skill by name. Returns the skill's instructions wrapped in <activated_skill> tags. Use this when you identify a task that matches a skill's description.",
  readOnly: true,
  input_schema: {
    type: "object",
    properties: {
      name: {
        type: "string",
        description: "The name of the skill to activate.",
      },
    },
    required: ["name"],
  },
  execute: async (
    args: Record<string, unknown>,
    _context?: ToolExecutionContext,
  ): Promise<ToolResult> => {
    const skillName = String(args.name);
    const skillBody = getSkillBody(skillName);

    if (!skillBody) {
      return { output: `Error: Skill '${skillName}' not found.` };
    }

    const output = `<activated_skill name="${skillName}">\n<instructions>\n${skillBody}\n</instructions>\n</activated_skill>`;
    return { output };
  },
};
register(activateSkillTool);

File: internal/tools/activate_skill.go
package tools

import (
	"context"
	"fmt"

	"minicode/internal/domain"
	"minicode/internal/skills"
)

// ActivateSkillTool provides instructions for a specific skill.
type ActivateSkillTool struct{}

// NewActivateSkillTool creates an ActivateSkillTool.
func NewActivateSkillTool() *ActivateSkillTool {
	return &ActivateSkillTool{}
}

func (t *ActivateSkillTool) Name() string { return "ActivateSkill" }
func (t *ActivateSkillTool) Description() string {
	return "Activate a specific skill and receive its detailed instructions. Use this when the user requests a task related to a listed skill."
}
func (t *ActivateSkillTool) ReadOnly() bool    { return true }
func (t *ActivateSkillTool) Interactive() bool { return false }
func (t *ActivateSkillTool) Requires() []ToolRequirement { return nil }

func (t *ActivateSkillTool) InputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"name": map[string]any{"type": "string", "description": "The name of the skill to activate"},
		},
		"required": []string{"name"},
	}
}

func (t *ActivateSkillTool) Execute(ctx context.Context, args map[string]any, tc ToolContext) (domain.ToolResult, error) {
	name, _ := args["name"].(string)
	if name == "" {
		tc.Log("ActivateSkill: name is required")
		return domain.ToolResult{Output: "Error: name is required"}, nil
	}

	tc.Log("ActivateSkill: activating skill", "name", name)

	body := skills.Body(name)
	if body == "" {
		tc.LogErr("ActivateSkill: skill not found", "name", name)
		return domain.ToolResult{Output: fmt.Sprintf("Error: Skill '%s' not found.", name)}, nil
	}

	tc.Log("ActivateSkill: skill activated", "name", name, "body_length", len(body))
	return domain.ToolResult{
		Output: fmt.Sprintf("<activated_skill name=\"%s\">\n<instructions>\n%s\n</instructions>\n</activated_skill>", name, body),
	}, nil
}

func init() { Register(NewActivateSkillTool()) }
